### PR TITLE
Init mixed precision training interface

### PR DESCRIPTION
**Pull request by kuke**
Simple Usage：

1. Cast specific inputs to float16 (Q: Should we do it automatically?)
```
  imgs = fluid.layers.cast(images, "float16")
```

2. Decorate optimizer and minimize scaled loss

```
 optimizer = fluid.optimizer.Adam(learning_rate=0.001)
	
 mp_optimizer = fluid.contrib.mixed_precision.decorate(
	            optimizer=optimizer, init_loss_scaling=8.0)
	
 scaled_loss, _, _ = mp_optimizer.minimize(avg_cost)
```

See [test_image_classification_fp16.py](https://github.com/PaddlePaddle/Paddle/blob/71c14fda6e0f027e67cee4bfcadc4b675024bddb/python/paddle/fluid/contrib/tests/test_image_classification_fp16.py) for details.

### CODE REVIEW

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+from ... import layers
+from ... import unique_name
+from . import fp16_utils
+from .fp16_utils import create_master_params_grads, master_param_to_train_param
+
+__all__ = ["decorate"]
+
+
+class OptimizerWithMixedPrecison(object):
+    def __init__(self, optimizer, init_loss_scaling, use_dynamic_loss_scaling):
```
Comment (qingqing01) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py:40]: Need comments for input arguments. Same as below.

File: python/paddle/fluid/initializer.py [hunk 2]
```diff
@@ -482,14 +482,28 @@ def __call__(self, var, block):
         if self._seed == 0:
             self._seed = block.program.random_seed
 
+# to be compatible of fp16 initalizers
+        if var.dtype == VarDesc.VarType.FP16:
+            out_dtype = VarDesc.VarType.FP32
+            out_var = block.create_var(
+                name=unique_name.generate(".".join(
+                    ['truncated_gaussian_random', 'tmp'])),
```
Comment (qingqing01) [inline python/paddle/fluid/initializer.py]: Maybe  use `var.name` as part of prefix for `unique_name.generate` a new name.

File: python/paddle/fluid/initializer.py [hunk 3]
```diff
@@ -500,15 +514,24 @@ def __call__(self, var, block):
...
                 },
                 stop_gradient=True)
+
+        if var.dtype == VarDesc.VarType.FP16:
+            block.append_op(
+                type="cast",
+                inputs={"X": out_var},
+                outputs={"Out": var},
+                attrs={"in_dtype": out_var.dtype,
+                       "out_dtype": var.dtype})
```
Comment (qingqing01) [inline python/paddle/fluid/initializer.py:559]: Wether the other initializer needs to do this FP32 weight creating and casting?

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        type="cast",
+        inputs={"X": i},
+        outputs={"Out": o},
+        attrs={
+            "in_dtype": core.VarDesc.VarType.FP32,
+            "out_dtype": core.VarDesc.VarType.FP16
+        })
+
+
+def copy_to_master_param(p, block):
```
Comment (qingqing01) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:39]: Need comments for all tehse APIs.

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        name=v.name + ".master")
+    return new_p
+
+
+def create_master_params_grads(params_grads, main_prog, startup_prog,
+                               loss_scaling):
+    master_params_grads = []
+    tmp_role = main_prog._current_role
+    OpRole = core.op_proto_and_checker_maker.OpRole
+    main_prog._current_role = OpRole.Backward
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: You can use `_backward_role_guard` here.

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def copy_to_master_param(p, block):
+    v = block.vars.get(p.name, None)
+    if v is None:
+        raise ValueError("no param name %s found!" % p.name)
+    new_p = framework.Parameter(
+        block=block,
+        shape=v.shape,
+        dtype=core.VarDesc.VarType.FP32,
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:54]: Why the dtype must be `FP32` here? Is it possible that the dtype is `fp64`?

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    tmp_role = main_prog._current_role
+    OpRole = core.op_proto_and_checker_maker.OpRole
+    main_prog._current_role = OpRole.Backward
+    for p, g in params_grads:
+        # create master parameters
+        master_param = copy_to_master_param(p, main_prog.global_block())
+        startup_master_param = startup_prog.global_block()._clone_variable(
+            master_param)
+        startup_p = startup_prog.global_block().var(p.name)
+        cast_fp16_to_fp32(startup_p, startup_master_param, startup_prog)
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: You should check that `p.type` is not `fp32`.

Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: You should check that `p.type` is not `fp32` here.

File: python/paddle/fluid/contrib/mixed_precision/decorator.py (hunk 1 above)
Comment (kuke) [reply python/paddle/fluid/contrib/mixed_precision/decorator.py:40]: Done

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 4 above)
Comment (kuke) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:39]: Done

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 5 above)
Comment (kuke) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: Done

File: python/paddle/fluid/initializer.py (hunk 2 above)
Comment (kuke) [reply python/paddle/fluid/initializer.py]: Done

File: python/paddle/fluid/initializer.py (hunk 3 above)
Comment (kuke) [reply python/paddle/fluid/initializer.py:559]: Yes. They all need. And all initializers have been made to support float16.

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 7 above)
Comment (kuke) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: Unified the two cast functions.

Comment (kuke) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py]: The same as above

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 6 above)
Comment (kuke) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:54]: In principle, the master parameter can be `float64`. But there are some [hard-coded implementations](https://github.com/PaddlePaddle/Paddle/blob/1202d3fc740b065590f93a6029ab2126ca9dafff/python/paddle/fluid/layers/nn.py#L3038), and the `fp64` support seems not to be that straightforward. So we are going to only support `float32` temporarily because it is more common used. Maybe we can go back to consider `fp64` some day in the future.

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        for p, g in params_grads:
+            # create master parameters
+            master_param = copy_to_master_param(p, main_prog.global_block())
+            startup_master_param = startup_prog.global_block()._clone_variable(
+                master_param)
+            startup_p = startup_prog.global_block().var(p.name)
+            # fp16 -> fp32
+            append_cast_op(startup_p, startup_master_param, startup_prog)
+            # cast fp16 gradients to fp32 before apply gradients
+            if g.name.find("batch_norm") > -1:
```
Comment (gongweibao) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:93]: Can we add attribute to these operator desc instead of hard code?

File: paddle/fluid/API.spec [hunk 9]
```diff
@@ -424,6 +424,7 @@ paddle.fluid.contrib.HDFSClient.upload (ArgSpec(args=['self', 'hdfs_path', 'loca
 paddle.fluid.contrib.multi_download (ArgSpec(args=['client', 'hdfs_path', 'local_path', 'trainer_id', 'trainers', 'multi_processes'], varargs=None, keywords=None, defaults=(5,)), ('document', '100927be598ed8f9eaa1f3ef1b23568a'))
 paddle.fluid.contrib.multi_upload (ArgSpec(args=['client', 'hdfs_path', 'local_path', 'multi_processes', 'overwrite', 'sync'], varargs=None, keywords=None, defaults=(5, False, True)), ('document', '183f34c83d30dbe16e09e8716c41958a'))
 paddle.fluid.contrib.extend_with_decoupled_weight_decay (ArgSpec(args=['base_optimizer'], varargs=None, keywords=None, defaults=None), ('document', 'a1095dfd4ec725747f662d69cd7659d4'))
+paddle.fluid.contrib.decorate (ArgSpec(args=['optimizer', 'init_loss_scaling', 'use_dynamic_loss_scaling'], varargs=None, keywords=None, defaults=(1.0, False)), ('document', '089f0c8d7c03bd3d0edc3ac83dbe41fd'))
```
Comment (chengduoZH) [inline paddle/fluid/API.spec]: The name(`decorate`) is not explicit.

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+from ... import unique_name
+from . import fp16_utils
+from .fp16_utils import create_master_params_grads, master_param_to_train_param
+
+__all__ = ["decorate"]
+
+
+class OptimizerWithMixedPrecison(object):
+    """
+    Optimizer class with mixed-precision training.
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py]: You should introduce the implementation of `OptimizerWithMixedPrecison` detailly and give an example here.

File: python/paddle/fluid/contrib/tests/test_image_classification_fp16.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
+#   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (gongweibao) [inline python/paddle/fluid/contrib/tests/test_image_classification_fp16.py:1]: This file is duplicated?
Can add arguments to the exists unittest instead of a new one?

File: python/paddle/fluid/initializer.py [hunk 12]
```diff
@@ -154,17 +154,41 @@ def __call__(self, var, block):
         """
         assert isinstance(var, framework.Variable)
         assert isinstance(block, framework.Block)
+
+        # to be compatible of fp16 initializers
+        if var.dtype == VarDesc.VarType.FP16:
+            out_dtype = VarDesc.VarType.FP32
+            out_var = block.create_var(
+                name=unique_name.generate(".".join(
+                    ['constant_init', var.name, 'tmp'])),
```
Comment (gongweibao) [inline python/paddle/fluid/initializer.py:163]: Where are these vars used?

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/initializer.py (hunk 12 above)
Comment (kuke) [reply python/paddle/fluid/initializer.py:163]: In the startup program

File: python/paddle/fluid/contrib/tests/test_image_classification_fp16.py (hunk 11 above)
Comment (kuke) [reply python/paddle/fluid/contrib/tests/test_image_classification_fp16.py:1]: 'Cause this interface is in `contrib`, we'd better use a seperate unit test file

File: paddle/fluid/API.spec (hunk 9 above)
Comment (kuke) [reply paddle/fluid/API.spec]: Fixed

File: python/paddle/fluid/contrib/mixed_precision/decorator.py (hunk 10 above)
Comment (kuke) [reply python/paddle/fluid/contrib/mixed_precision/decorator.py]: Added more details

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 8 above)
Comment (kuke) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:93]: Next step, wo can optimize such hard code

**Review by gongweibao — APPROVED** [review]

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    """
+    Optimizer with mixed-precision (MP) training. This is a wrapper of a common 
+    optimizer, plus the support of mixed-precision pretraining. The object
+    of this class almost has the same behavior as the common optimizer, with the 
+    methods `minimize()`, `backward()`, `apply_gradients()` implemented. 
+    Additionally, it enables the MP training automatically, i.e, the creation 
+    and maintenance of master parameters, scaling of loss, etc.
+
+    Args:
+        optimizer (Optimizer): A common Optimizer object.
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py:35]: -> `A Optimizer object.`

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+        self._optimizer = optimizer
+        self._param_grads = None
+        self._train_program = default_main_program()
+        self._startup_prog = default_startup_program()
+        self._loss_scaling = init_loss_scaling
+        self._use_dynamic_loss_scaling = use_dynamic_loss_scaling
+
+        # Ensure the data type of learning rate vars is float32 (same as the 
+        # master parameter dtype)
+        if isinstance(optimizer._learning_rate, float):
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py:50]: What will happen when the optimizer._learning_rate is not `float`?

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    and maintenance of master parameters, scaling of loss, etc.
+
+    Args:
+        optimizer (Optimizer): A common Optimizer object.
+        init_loss_scaling (float): The initial loss scaling factor.
+        use_dynamic_loss_scaling (bool): Whether to use dynamic loss scaling.
+    """
+
+    def __init__(self, optimizer, init_loss_scaling, use_dynamic_loss_scaling):
+        self._optimizer = optimizer
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py:41]: Please check the type of `optimizer`.

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 8 above)
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:93]: Agreed with @gongweibao, maybe you should get the op that generates `g`.

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+            The scaled loss by scaling factor, the list of optimize ops, and a
+            list of master parameters and gradients.
+        """
+        master_params_grads, scaled_loss = self.backward(loss)
+        optimize_ops = self.apply_gradients(master_params_grads)
+
+        return scaled_loss, optimize_ops, master_params_grads
+
+
+def decorate(optimizer, init_loss_scaling=1.0, use_dynamic_loss_scaling=False):
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py:129]: How about `extend_optimizer_with_mixed_precison`?

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    """ 
+    Decorate the given optimizer to adapt to the mixed-precision training.
+
+    Args:
+        optimizer(Optimizer): A common Optimizer.
+        init_loss_scaling(float): The initial loss scaling factor.
+        use_dynamic_loss_scaling(bool): Whether to use dynamic loss scaling.
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py:136]: Please give the default value.

File: python/paddle/fluid/contrib/mixed_precision/decorator.py [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+            startup_program (Program|None): The startup Program for initializing 
+                                       parameters in `parameter_list`.
+            parameter_list (list|None): A list of Variables to update.
+            no_grad_set (set|None): A set of Variables should be ignored.
+            callbacks (list|None): A list of callables to run when appending 
+                                   backward operator for one parameter.
+
+        Returns:
+            A list of (param, grad), which is a tuple of a parameter and its 
+            gradient respectively, and the scaled loss.
```
Comment (chengduoZH) [inline python/paddle/fluid/contrib/mixed_precision/decorator.py:84]: Please explain the `master_params`?

**Review by chengduoZH — COMMENTED** [review]
The code should be polished.

**Review by chengduoZH — APPROVED** [review]
The code should be polished.

**Review by shanyi15 — APPROVED** [review]
approve for under contrib menu
